Remove versionType from translog

dnhatn opened the pull request:
With the introduction of sequence number, we no longer use versionType to
resolve out of order collision in replication and recovery requests.

This PR removes removes the versionType from translog. We can only remove 
it in 7.0 because it is still required in a mixed cluster between 6.x and 5.x.

elasticmachine commented:
Pinging @elastic/es-distributed

bleskes commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -649,10 +649,10 @@ private IndexShardState changeState(IndexShardState newState, String reason) {
+                                                           boolean isRetry, SourceToParse sourceToParse)
         throws IOException {
-        return applyIndexOperation(seqNo, primaryTerm, version, versionType, autoGeneratedTimeStamp, isRetry,
+        return applyIndexOperation(seqNo, primaryTerm, version, VersionType.EXTERNAL, autoGeneratedTimeStamp, isRetry,
```
maybe it's better to pass null instead of VersionType.EXTERNAL? this way people may think it's the original one?

bleskes replied:
if we do that, maybe assert in the Engine.Op constructor that if the origin is not a primary, version type is null?

ywelsch replied:
+1 on this suggestion

bleskes commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -740,9 +740,8 @@ private IndexShardState changeState(IndexShardState newState, String reason) {
-                                                             VersionType versionType) throws IOException {
-        return applyDeleteOperation(seqNo, primaryTerm, version, type, id, versionType, Engine.Operation.Origin.REPLICA);
+    public Engine.DeleteResult applyDeleteOperationOnReplica(long seqNo, long version, String type, String id) throws IOException {
+        return applyDeleteOperation(seqNo, primaryTerm, version, type, id, VersionType.EXTERNAL, Engine.Operation.Origin.REPLICA);
```
some comment as before

ywelsch commented on server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java:
```diff
@@ -202,9 +203,27 @@ private synchronized boolean assertNoSeqNumberConflict(long seqNo, BytesReferenc
-                        "prvOp [" + prvOp + "], newOp [" + newOp + "]", previous.v2());
+                // we need to exclude versionType from this check because it's removed in 7.0
+                final boolean sameOp;
+                if (prvOp instanceof Translog.Index && newOp instanceof Translog.Index) {
```
versionType does not exist anymore in `Translog.Index` and `Translog.Delete`, so can't we just use `equals` on the `Translog.Operation` that we recreated from the bytes?

dnhatn commented:
@bleskes and @ywelsch I have addressed your comments and pushed  
https://github.com/elastic/elasticsearch/pull/31945/commits/d3780eeee5cd35121ced39bf243e8e1d46fe7ef3 to reserve some format versions for 6.x. Can you please take another look?

ywelsch commented on server/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -1168,6 +1168,8 @@ public long startTime() {
         public Index(Term uid, ParsedDocument doc, long seqNo, long primaryTerm, long version, VersionType versionType, Origin origin,
                      long startTime, long autoGeneratedIdTimestamp, boolean isRetry) {
             super(uid, seqNo, primaryTerm, version, versionType, origin, startTime);
+            assert (origin == Origin.PRIMARY && versionType != null) || (origin != Origin.PRIMARY && versionType == null) :
```
simpler to write `(origin == Origin.PRIMARY) == (versionType != null)`

ywelsch commented on server/src/main/java/org/elasticsearch/index/translog/Translog.java:
```diff
@@ -1010,16 +1009,17 @@ public Source(BytesReference source, String routing) {
         public static final int FORMAT_6_0 = 8; // since 6.0.0
-        public static final int FORMAT_NO_PARENT = FORMAT_6_0 + 1; // since 7.0
-        public static final int SERIALIZATION_FORMAT = FORMAT_NO_PARENT;
+        public static final int FORMAT_7_0 = FORMAT_6_0 + 5; // reserve some versions for 6.x
```
I don't think we should preemptively do this. If it comes to us requiring the extra versions for a change in 6.x, and we have not released 7.0, we can up the format version for the current change. If we have released 7.0 at that point, we cannot make any changes to the format anyhow.

bleskes replied:
+1

dnhatn replied:
+1

bleskes commented:
+1

On Tue, 17 Jul 2018 at 9:52 AM, Yannick Welsch <notifications@github.com>
wrote:

> *@ywelsch* approved this pull request.
>
> LGTM
> ------------------------------
>
> In server/src/main/java/org/elasticsearch/index/engine/Engine.java
> <https://github.com/elastic/elasticsearch/pull/31945#discussion_r202914053>
> :
>
> > @@ -1168,6 +1168,8 @@ public long startTime() {
>          public Index(Term uid, ParsedDocument doc, long seqNo, long primaryTerm, long version, VersionType versionType, Origin origin,
>                       long startTime, long autoGeneratedIdTimestamp, boolean isRetry) {
>              super(uid, seqNo, primaryTerm, version, versionType, origin, startTime);
> +            assert (origin == Origin.PRIMARY && versionType != null) || (origin != Origin.PRIMARY && versionType == null) :
>
> simpler to write (origin == Origin.PRIMARY) == (versionType != null)
> ------------------------------
>
> In server/src/main/java/org/elasticsearch/index/translog/Translog.java
> <https://github.com/elastic/elasticsearch/pull/31945#discussion_r202923141>
> :
>
> > @@ -1010,16 +1009,17 @@ public Source(BytesReference source, String routing) {
>      public static class Index implements Operation {
>
>          public static final int FORMAT_6_0 = 8; // since 6.0.0
> -        public static final int FORMAT_NO_PARENT = FORMAT_6_0 + 1; // since 7.0
> -        public static final int SERIALIZATION_FORMAT = FORMAT_NO_PARENT;
> +        public static final int FORMAT_7_0 = FORMAT_6_0 + 5; // reserve some versions for 6.x
>
> I don't think we should preemptively do this. If it comes to us requiring
> the extra versions for a change in 6.x, and we have not released 7.0, we
> can up the format version for the current change. If we have released 7.0
> at that point, we cannot make any changes to the format anyhow.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/31945#pullrequestreview-137698814>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA9bJ_QZ6KVFJzqD2H6McV0_Bd0u6zkEks5uHZfJgaJpZM4VKVI9>
> .
>

dnhatn commented:
Thanks @bleskes and @ywelsch for reviewing!
